cylc/flow/config.py

```diff
@@ -305,7 +305,7 @@ def __init__(
         self.mem_log("config.py: before inheritance")
         self.compute_inheritance()
         self.mem_log("config.py: after inheritance")
-
+        self.cfg = host_to_platform_upgrader(self.cfg)
```

Comment:
Here your doing the upgrade post-inheritance, so this flow:
```
[runtime]
    [[root]]
          [[[remote]]]
                host = cheese
    [[foo<m>]]  # m = 1.. 1000
```
will result in 1001 upgrades (root and foo_m) instead of 1 (just root) - and 1001 validation warnings!

We have done every other syntax upgrade in the cfgspec module on the sparse config, after parsing and before inheritance processing, and then the upgraded result gets inherited through the family hierarchy as normal.

I guess you can't do that here because the upgrader assumes `localhost` for any namespace with no host or platform specified.  Pre-inheritance, most namespaces are in this category, but they aren't targeting localhost because they inherit the remote host.

This method also pulls every task (via `platform = localhost`) into the sparse config, which is only supposed to hold config items explicitly specified by the user (try `cylc get-config --sparse -i runtime ...`).

So I guess the question is, does platforms upgrade localhost selection prevent us from doing this the usual efficient way?  What about just inserting `platform = localhost` in the root namespace, if no platform or remote is specified there already?

Reply:
Thank goodness - I thought it should not be necessary.

Reply:
> What about just inserting platform = localhost in the root namespace, if no platform or remote is specified there already?

Jargon Brainfreeze - are we talking about setting a default in the config thus: 
```
with Conf('platform', VDR.V_STRING, 'localhost'):
```

Reply:
Maybe just add the default value at line 493 (no `with` needed) ... I think.  Then adapt your upgrader function to deal with that change.  (Not 100% sure though, since Oliver's fancy new self-documenting `with Conf(...)` refactor ... would need to have a play with it to confirm...)

Reply:
This looks like it's not going to work, because if I have the following suiterc runtime:
```
[[FAMILY]]
  [[[remote]]]
    host = supercomputer_login_node01
[[mytask]]
  inherit = FAMILY
  [[[job]]]
    batch system = slurm
```
and platforms thus:
```
[[supercomputer_slurm]]
hosts = supercomputer_login_node01, supercomputer_login_node02
batch system = slurm

[[smaller_spice_cluster]]
batch system = slurm
hosts = spice01, spice02
```
You'll end up with mytask being assigned `platform = smaller_spice_cluster` if you upgrade before inheritance, but `platform = supercomputer_slurm` if you do it after. 

Worse still, in this scenario FAMILY will fail if you try to upgrade at the normal time because there is no platform with `batch system = background` defined.

@dpmatthews  has suggested that we could put a warning about old syntax on upgrade and do the upgrading quietly where it is...

How much effort should we put into code which ought to be deprecated at Cylc9?

Reply:
Because the upgrader requires multiple pieces of information and because these could come from anywhere in the inheritance hierarchy I don't think we can upgrade families, only tasks.

Reply:
After a conversation with @dpmatthews and @oliver-sanders  I think we've concluded that because the Cylc 7 to platform upgrade is not a 1:1 mapping, but takes account of _all_ the items in the task `[remote]` and `[job]` sections we don't think that a traditional upgrade (i.e. before config expansion and inheritance is calculated) is possible. We've agreed that we think the best approach is to separate the deprecation warnings from the upgrade:

At suite validation the user will be warned that combinations of `host`, `batch system` and `batch submit command template` will be used to select platforms and will be deprecated at Cylc9. Further down the line this should include links to upgrade documentation.

Upgrade logic will be used on the fly by adding an interface to `platforms.py`:
```python
# Psuedocode really
get_platform(job_config):
  if job_config has platform name and not forbidden cylc 7 settings:
    get_platform_from_name()
  if job_config has platform name and forbidden with platforms cylc 7 settings:
    fail horribly
  if job_config not have platform name and forbidden with platforms cylc 7 settings:
    get_platform_from_cylc_7_settings()
  else:
    return localhost platform config
```

Reply:
OK, damn it, right.  Can we perhaps give a stronger-than-normal deprecation warning given the unusual complexity of this upgrade process, which could be quite slow in a huge suite? (and check no significant impact on workflows that don't need upgrading).

Reply:
> stronger-than-normal deprecation warning

Yes